### PR TITLE
Add Roslyn language service factories for our exported types.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- Verified this all worked via our Razor extension:
image

Wasn't going to check in the extension changes due to their lack of usefulness but if you feel different let me know.

#1260

### CODE REVIEW

File: src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultTagHelperResolverFactory.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+using System.Composition;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.VisualStudio.Shell;
+
+namespace Microsoft.VisualStudio.LanguageServices.Razor
+{
+    [ExportLanguageServiceFactory(typeof(ITagHelperResolver), RazorLanguage.Name, ServiceLayer.Default)]
+    internal class DefaultTagHelperResolverFactory : ILanguageServiceFactory
```
Comment (NTaylorMullen) [inline src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultTagHelperResolverFactory.cs]: We should really rename this TagHelper resolver to be a `HostTagHelperResolver` or `VisualStudioTagHelperResolver`. If you agree I can push that in separately.

Comment (rynowak) [inline src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultTagHelperResolverFactory.cs]: Sure

File: src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultRazorSyntaxFactsServiceFactory.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Razor;
+
+namespace Microsoft.VisualStudio.LanguageServices.Razor
+{
+    [ExportLanguageServiceFactory(typeof(RazorSyntaxFactsService), RazorLanguage.Name, ServiceLayer.Default)]
+    internal class DefaultRazorSyntaxFactsServiceFactory : ILanguageServiceFactory
```
Comment (rynowak) [inline src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultRazorSyntaxFactsServiceFactory.cs]: These should live in the Razor.Workspaces assembly

Comment (rynowak) [inline src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultRazorSyntaxFactsServiceFactory.cs]: Anything in ServiceLayer.Default should live there

Comment (NTaylorMullen) [inline src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultRazorSyntaxFactsServiceFactory.cs]: Ah, will have to move a lot of codez. Coolio, will do.

File: src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultTagHelperResolverFactory.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultTagHelperResolverFactory.cs]: This looks like the only one that needs to live in the VS assembly.

File: src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultTagHelperResolverFactory.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    {
+        private readonly VisualStudioWorkspace _workspace;
+        private readonly SVsServiceProvider _services;
+
+        [ImportingConstructor]
+        public DefaultTagHelperResolverFactory(VisualStudioWorkspace workspace, SVsServiceProvider services)
```
Comment (rynowak) [inline src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultTagHelperResolverFactory.cs]: For improved do-thingability you can use `[Export(typeof(SVsServiceProvider))] IServiceProvider services`. The `SVsServiceProvider` is just a marker.

File: src/Microsoft.VisualStudio.LanguageServices.Razor/ITagHelperResolver.cs [hunk 4]
```diff
@@ -4,11 +4,12 @@
 using System.Collections.Generic;
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Host;
 using Microsoft.CodeAnalysis.Razor;
 
 namespace Microsoft.VisualStudio.LanguageServices.Razor
 {
-    public interface ITagHelperResolver
+    public interface ITagHelperResolver : ILanguageService
```
Comment (rynowak) [inline src/Microsoft.VisualStudio.LanguageServices.Razor/ITagHelperResolver.cs]: Can we make a new interface for this in `Microsoft.CodeAnalysis.Razor` namespace (`Microsoft.AspNetCore.Razor.Workspaces` assembly).

I don't want to roslynize this with the wrong API

Comment (NTaylorMullen) [inline src/Microsoft.VisualStudio.LanguageServices.Razor/ITagHelperResolver.cs]: Sure.

File: src/Microsoft.VisualStudio.LanguageServices.Razor/RazorSyntaxFactsService.cs [hunk 5]
```diff
@@ -3,11 +3,12 @@
 
 using System.Collections.Generic;
 using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.CodeAnalysis.Host;
 using Microsoft.VisualStudio.Text;
 
 namespace Microsoft.VisualStudio.LanguageServices.Razor
 {
-    public abstract class RazorSyntaxFactsService
+    public abstract class RazorSyntaxFactsService : ILanguageService
```
Comment (rynowak) [inline src/Microsoft.VisualStudio.LanguageServices.Razor/RazorSyntaxFactsService.cs]: My general comment here, I don't really want 'roslynize' the wrong API. For instance this is in the wrong namespace/assembly. I think the simplest thing is to copy these base classes

**Comment by rynowak** [comment]
It would be useful to add something that would display the taghelpers as a POC but I understand if you want me to use my WPF-Fu on that instead

**Comment by NTaylorMullen** [comment]
Closing till we revisit this area.

**Pull request closed without merging**